### PR TITLE
Prometheus: Migrate range & instant Explore queries to backend

### PR DESCRIPTION
**Pull request by ivanahuckova**
This PR was closed and split into smaller PRs.

This PR is part of https://github.com/grafana/grafana/issues/37784. It tackles 2 points:
 - [x] Running of range queries in Explore trough backend @ivanahuckova
 - [x] Running of instant queries in Explore trough backend @ivanahuckova
 
 The Running of exemplars queries in Explore trough backend come next, but is not part of this PR.

What I did:

**Frontend**
1. I have fixed initial values for `stepMode, instant, range` in `PromExploreQueryEditor` as there were no initial values.
1. In `datasource.ts`, I have created new `prepareOptionsV2` method that prepares targets for backend queries. Currently it holds logic only for `Explore + range/instant queries` as other queries are not run trough this. This method will grow as we expand range of queries run trough backend and should include all `prepareTargets` for all targets. When we run all queries trough backend, this will be renamed to `prepareTargets` and current `prepareTargets` will be removed as not needed. It includes `processIntervalV2` method that calculates interval, rate interval and adds interval and rate interval to scopeVars. We have discussed if interval processing should be on frontend or backend, but because we are processing template variables on frontend, this need to be done on frontend as well, because we have `$__interval` and `$__rate_interval` variables that needs to be calculated for each query. 
1. Created new PromQuery param - `stepMS` which holds value of calculated step in ms. On backend, if we receive `stepMS` value, we know that we can just pass it and we don't have to calculate it on backend. 
1. In `datasource.ts`, I have added `applyTemplateVariables` and `filterQuery` methods that are used when running queries trough backend.
1. In `datasource.ts`, I have updated `query` method and added condition to run `Explore + range/instant queries` trough backend. These queries have then different processing.  
1. In `result_transformer.ts`, I have created `transformV2` to transform results of queries. The results are in dataFrame structure, however for instant queries (an in the future other queries), we want to use **table dataFrame** and therefore we need to run it trough `transformDFoTable` function that updatess dataFrame to include label as fields. 

**Backend**
1. Split backend into multiple files with test files:
- prometheus.go 
- query_parser.go 
- response_parser.go 
1. Created `ParseIntervalStringToTimeDuration` that has code that changes calculated interval string to time.druation.
1. Added parsing of all types of responses, not just matrix.

**How to manually test this**
https://gist.github.com/ivanahuckova/e717457b32265f1b4e0583f0861cbb27

### CODE REVIEW

File: public/app/plugins/datasource/prometheus/datasource.ts [hunk 1]
```diff
@@ -486,12 +553,48 @@ export class PrometheusDatasource extends DataSourceApi<PromQuery, PromOptions>
     return { __rate_interval: { text: rateInterval + 's', value: rateInterval + 's' } };
   }
 
+  processIntervalV2(target: PromQuery, options: DataQueryRequest<PromQuery>) {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/prometheus/datasource.ts:556]: This is the same logic as is in [createQuery](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/datasource.ts#L407-L451)

**Review by dsotirakis — COMMENTED** [review]
This is 🔥  @ivanahuckova!
Added some comments!

File: pkg/tsdb/prometheus/query_parser.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	RangeQuery     bool   `json:"range"`
+	InstantQuery   bool   `json:"instant"`
+	IntervalFactor int    `json:"intervalFactor"`
+	// IsBackendQuery is used to track if we need to calculate interval or if it was calculated
+	// on client side. We can remove this when we fully migrate to running all queries trough backend.
+	IsBackendQuery bool `json:"isBackendQuery"`
+}
+
+func (s *Service) parseQuery(dsInfo *DatasourceInfo, queryContext *backend.QueryDataRequest) (
+	[]*PrometheusQuery, error) {
```
[the remark this answers is not in the document]
Comment (dsotirakis) [reply pkg/tsdb/prometheus/query_parser.go]: Yep it does! 😁

**Comment by ivanahuckova** [comment]
Based on conversation with @gabor I have added 1 change. We are going to create `stepMS` query parameter that is going to hold the value of calculated step in ms as number. If backend receives this step it knows that step was calculated and it can just pass it to final query. Otherwise we have to calculate it on backend. As mentioned in this PR description, the goal is to minimize/remove calculation of step on backend and keep it just on frontend.

**Comment by ivanahuckova** [comment]
@Elfo404 @gabor @dsotirakis could you please review again. 🙂

File: pkg/tsdb/prometheus/query_parser.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	qs := []PrometheusQuery{}
+	for _, query := range queryContext.Queries {
+		model := &QueryModel{}
+		err := json.Unmarshal(query.JSON, model)
+		if err != nil {
+			return nil, err
+		}
+
+		step := time.Duration(model.StepMS) * time.Millisecond
+		if step == time.Duration(0) {
```
Comment (dsotirakis) [inline pkg/tsdb/prometheus/query_parser.go:34]:
```suggestion
		if model.StepMS == 0 {
```
nit, but maybe that would be more clear?

File: pkg/tsdb/prometheus/query_parser.go [hunk 4]
```diff
@@ -78,7 +71,7 @@ func createStep(dsInfo *DatasourceInfo, model *QueryModel, query backend.DataQue
 		return time.Duration(0), err
 	}
 
-	if hasQueryInterval {
+	if hasQueryInterval && model.StepMode != "" {
```
Comment (dsotirakis) [inline pkg/tsdb/prometheus/query_parser.go:74]: Did the changes to add `model.StepMS` affected this piece as well, or is it something we missed before?

**Pull request closed without merging**